src/System.Runtime.Extensions/src/System/Net/WebUtility.cs

```diff
@@ -357,6 +357,66 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
             return expandedBytes;
         }
 
+        private static unsafe string UrlEncodeInternal(string originalString)
+        {
+            fixed (char* pch = originalString)
+            {
+                int safeCharCount = 0;
+                int spaceCharCount = 0;
+                for (int i = 0; i < originalString.Length; i++)
+                {
+                    char c = pch[i];
+                    if (c == ' ')
+                    {
+                        spaceCharCount++;
+                    }
+                    else if (IsUrlSafeChar(c))
+                    {
+                        safeCharCount++;
+                    }
+                }
+
+                if (spaceCharCount == 0 && safeCharCount == originalString.Length)
+                {
+                    // Nothing to expand
+                    return originalString;
+                }
+                
+                int unsafeByteCount = Encoding.UTF8.GetByteCount(pch, originalString.Length) - safeCharCount - spaceCharCount;
```

Comment:
Oh right - I don't have a PC on hand right now, but I seem to recall that this doesn't work for surrogate pairs because the length is 1? Is this right